cg/meta/transfer/external_data.py

```diff
@@ -0,0 +1,211 @@
+        all_fastqs: List[Path] = []
+        for leaf in sample_folder.iterdir():
+            abs_path: Path = sample_folder.joinpath(leaf)
+            LOG.info("Found file %s inside folder %s" % (str(abs_path), sample_folder))
+            if abs_path.suffixes == [".fastq", ".gz"]:
+                all_fastqs.append(abs_path)
+        return all_fastqs
+
+    def get_all_paths(self, lims_sample_id: str, customer: str) -> List[Path]:
+        """Returns the paths of all fastq files associated to the sample."""
+        fastq_folder: Path = self.create_destination_path(
+            lims_sample_id=lims_sample_id, customer=customer, raw_path=self.hasta_path
+        )
+        all_fastq_in_folder: List[Path] = self.get_all_fastq(sample_folder=fastq_folder)
+        return all_fastq_in_folder
+
+    def create_hk_bundle(
+        self, bundle_name: str, dry_run: bool, data_dict: dict
+    ) -> Optional[Tuple[Bundle, Version]]:
+        """Creates and returns a housekeeper bundle and the new version."""
+        if dry_run:
+            LOG.info(
+                "Would have created bundle %s in housekeeper, but this is dry-run", bundle_name
+            )
+            return
+        bundle_result: Tuple[Bundle, Version] = self.housekeeper_api.add_bundle(
+            bundle_data=data_dict
+        )
+        return bundle_result
+
+    def check_bundle_fastq_files(
+        self, fastq_paths: List[Path], lims_sample_id: str, last_version: Version
+    ) -> List[Path]:
+        """Checks if any of the fastq files are already added. Returns a list of files that should be added."""
+        for file in self.housekeeper_api.get_files(
+            bundle=lims_sample_id, tags=["fastq"], version=last_version.id
+        ):
+            if Path(file.path) in fastq_paths:
+                fastq_paths.remove(Path(file.path))
+                LOG.info(
+                    "Path %s is already linked to bundle %s in housekeeper"
+                    % (file.path, lims_sample_id)
+                )
+        return fastq_paths
+
+    def add_files_to_bundles(
+        self, fastq_paths: List[Path], last_version: Version, lims_sample_id: str
+    ) -> bool:
+        """Adds the given files to the the hk-bundle if the md5sum matches or no md5sum is provided."""
+        for path in fastq_paths:
+            do_commit: bool = True
+            if Path(str(path) + ".md5").exists():
+                given_md5sum: str = extract_md5sum(md5sum_file=Path(str(path) + ".md5"))
+                do_commit: bool = check_md5sum(file_path=path, md5sum=given_md5sum)
+            else:
+                LOG.info("No md5 file found for file  %s" % path + ".md5")
+            if do_commit:
+                LOG.info("Adding path %s to bundle %s in housekeeper" % (path, lims_sample_id))
+                self.housekeeper_api.add_file(path=path, version_obj=last_version, tags=["fastq"])
+            else:
+                return False
```

Comment:
```suggestion
            else:
                return False
```
Do you want to break in the middle of the for loop? You migt end up with some files added and some not added. Would it not be better to try to check all md5sum first and then commit all files if they are all ok or abort all and broadcast it to the user if anyone fails?

Reply:
How about splitting the functionality up in one function that checks and compares md5sums and one that does the actual commit. If the md5sum do not match for a path, the path is added to a list. The list is returned after looping through all. The next function checks the list for members and croaks and logs if there are any, otherwise happily proceeds to add all original paths to HK - knowing that they have all been verified. This is just a suggestion - I must confess I have not really dug deep into this code.

Reply:
Re-written according to your suggestion.